[PrintAsObjC] Print availability on classes, protocols, and categories

jrose-apple opened the pull request:
Now that Clang has availability diagnostics too (-Wpartial-availability and friends) we need to get this correct, or people will get warnings in the generated header!

Also fix up an issue with availability messages being omitted in certain cases.

rdar://problem/33313703

jrose-apple committed:
[PrintAsObjC] Print availability on classes, protocols, and categories

Now that Clang has availability diagnostics too
(-Wpartial-availability and friends) we need to get this correct, or
people will get warnings in the generated header!

rdar://problem/33313703

jrose-apple committed:
[PrintAsObjC] Simplify loop over availability attributes.

No functionality change. Diff best viewed ignoring whitespace.

jrose-apple committed:
[PrintAsObjC] Include messages for platform always-unavailable decls.

A misplaced brace left the messages in empty space. Noticed by
inspection.

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux

jrose-apple commented:
I recommend reviewing this one commit-by-commit and/or with a diff that ignores whitespace: https://github.com/apple/swift/pull/11252/files?w=1

jrose-apple commented:
```
error: argument labels '(extendingOrTruncating:)' do not match any available overloads
        return Int(extendingOrTruncating: result)
               ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

@moiseev, looks like you missed SwiftPM in your update?

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux

jrose-apple commented:
Review ping for @DougGregor or @devincoughlin.

jrose-apple commented:
@swift-ci Please smoke test

DougGregor commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -710,132 +710,131 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+      case PlatformKind::watchOSApplicationExtension:
+        plat = "watchos_app_extension";
+        break;
+      default:
```
Can we not have a `default` case here? The effect is the same, but we'll get warnings if `PlatformKind` ever gets extended.

jrose-apple replied:
That's pre-existing code I just moved around, but sure.

DougGregor replied:
Ah, ok.

jrose-apple committed:
[PrintAsObjC] default -> fully-covered switch

Per feedback from Doug. The last case was unreachable anyway, so no
functionality change.

jrose-apple commented:
@swift-ci Please smoke test
